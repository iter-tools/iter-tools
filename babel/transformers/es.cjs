const { resolve } = require('path');
const babelJest = require('babel-jest');

module.exports = babelJest.createTransformer({
  configFile: resolve(__dirname, '../jest.config.cjs'),
});
