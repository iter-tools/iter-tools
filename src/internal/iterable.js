import { isIterable } from '../impls/is-iterable/is-iterable.js';
import { isLoopable } from '../impls/is-loopable/is-loopable.js';
import { isWrappable } from '../impls/is-wrappable/is-wrappable.js';
import { nullableWrap as wrap } from './wrap.js';
import { variadicCurryWithValidation } from './curry.js';
import { _, __iterate } from './symbols.js';

export { wrap, isIterable, isLoopable, isWrappable };

export function ensureIterable(value) {
  if (!isWrappable(value)) {
    if (typeof value.next === 'function') {
      throw new TypeError(
        'iter-tools received a value that looked like an iterator but was not iterable. Get help fixing this: https://github.com/iter-tools/iter-tools/wiki/Making-iterators-iterable',
      );
    } else throw new TypeError('Expected an iterable, null, or undefined');
  } else if (isIterable(value)) {
    return value;
  } else {
    return wrap(value);
  }
}

export function callReturn(iterator) {
  if ('return' in iterator) iterator.return();
}

export function BaseResultIterable(fn, args, iterablesArg) {
  this[_] = {
    fn,
    args,
    iterablesArg,
    staticIterator: null,
  };
}

Object.assign(BaseResultIterable.prototype, {
  constructor: BaseResultIterable,

  [__iterate]() {
    const { fn, iterablesArg, args } = this[_];
    return fn(iterablesArg, ...args);
  },

  next() {
    const this_ = this[_];
    this_.staticIterator = this_.staticIterator || this[__iterate]();
    return this_.staticIterator.next();
  },

  return(value) {
    const this_ = this[_];
    this_.staticIterator = this_.staticIterator || this[__iterate]();
    return this_.staticIterator.return(value);
  },

  throw(error) {
    const this_ = this[_];
    this_.staticIterator = this_.staticIterator || this[__iterate]();
    return this_.staticIterator.throw(error);
  },
});

export function ResultIterable(...args) {
  BaseResultIterable.apply(this, args);
}

ResultIterable.prototype = Object.assign(Object.create(BaseResultIterable.prototype), {
  constructor: ResultIterable,
  [Symbol.iterator]() {
    return this[__iterate]();
  },
});

function SimpleResultIterable(...args) {
  ResultIterable.apply(this, args);
}

SimpleResultIterable.prototype = Object.assign(Object.create(ResultIterable.prototype), {
  constructor: SimpleResultIterable,
  [__iterate]() {
    return this[_].fn(...this[_].args);
  },
});

function makeFunctionConfig(fn, fnConfig = {}) {
  const { validateArgs, variadic, reduces, optionalArgsAtEnd, minArgs, maxArgs } = fnConfig;

  return {
    fn,
    validateArgs: validateArgs || ((_) => {}),
    variadic: !!variadic,
    reduces: !!reduces,
    optionalArgsAtEnd: !!optionalArgsAtEnd,
    minArgs: minArgs === undefined ? fn.length - 1 : minArgs,
    maxArgs: maxArgs === undefined ? fn.length - 1 : maxArgs,
    isIterable: isWrappable,
    iterableType: 'iterable',
    applyOnIterableArgs: ensureIterable,
    IterableClass: ResultIterable,
  };
}

export function cache(it) {
  return [...it];
}

export function wrapWithResultIterable(fn, { validateArgs = (_) => _ } = {}) {
  return (...args) => {
    validateArgs(args);
    return new SimpleResultIterable(fn, args);
  };
}

export const iterableCurry = (fn, config) => {
  return variadicCurryWithValidation(makeFunctionConfig(fn, config));
};
