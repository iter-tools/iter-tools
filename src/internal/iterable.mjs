import { variadicCurryWithValidation } from './curry'
const emptyArr = []

export function isIterable (i) {
  return Boolean(i != null && i[Symbol.iterator])
}

export function ensureIterable (i) {
  if (i == null) {
    return emptyArr[Symbol.iterator]()
  } else if (!isIterable(i)) {
    if (typeof i.next === 'function') {
      throw new TypeError('Iterators are not supported arguments to iter-tools. You must wrap them using the `iterable` method.')
    }
    throw new TypeError('The argument is not an iterable or null')
  }
  return i
}

export function isValidIterableArgument (i) {
  return i == null || isIterable(i)
}

export const iterableCurry = (fn, { variadic = false, reduces = false } = {}, minArgs, maxArgs) => {
  return variadicCurryWithValidation(
    isValidIterableArgument,
    'iterable',
    ensureIterable,
    fn,
    variadic,
    reduces,
    false,
    minArgs,
    maxArgs
  )
}
