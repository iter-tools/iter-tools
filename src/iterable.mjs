const emptyArray = []

export default function iterable (iterator) {
  if (iterator == null) {
    return emptyArray
  } else if (iterator[Symbol.iterator]) {
    return iterator
  } else if (typeof iterator !== 'object' || typeof iterator.next !== 'function') {
    throw new Error(`Expected to receive an iterator of the form {next()}, but instead received: ${iterator}`)
  }

  return {
    [Symbol.iterator] () {
      return iterator
    }
  }
}
