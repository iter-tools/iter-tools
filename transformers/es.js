const createTransformer = require('./create-transformer');

module.exports = createTransformer('babel-es.config.js');
