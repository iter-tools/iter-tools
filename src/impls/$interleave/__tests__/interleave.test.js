/**
 * @generated-from ./$interleave.test.js
 * This file is autogenerated from a template. Please do not edit it directly.
 * To rebuild it from its template use the command
 * > npm run generate
 * More information can be found in CONTRIBUTING.md
 */

import { Iterable } from '../../../types/iterable.js';
import { interleave, Peekerator } from '@iter-tools/es';
import { wrap, unwrap, unwrapDeep, anyType } from '../../../test/helpers.js';

const roundRobinStrategy = function* (
  options: Record<string, any>,
  all: Peekerator<Peekerator<number>>,
  a: Peekerator<number>,
  b: Peekerator<number>,
  c: Peekerator<number>,
) {
  while (!all.done) {
    if (!a.done) {
      yield a.value;
      a.advance();
    }
    if (!b.done) {
      yield b.value;
      b.advance();
    }
    if (!c.done) {
      yield c.value;
      c.advance();
    }
  }
};

const roundRobin = interleave(roundRobinStrategy, {});

describe('interleave', () => {
  describe('when there are only empty sources', () => {
    it('yields no values', () => {
      expect(unwrap(roundRobin(null, undefined, []))).toEqual([]);
    });
  });

  describe('when all sources are the same size', () => {
    it('yields all values collated', () => {
      const sources = [wrap([1, 2, 3]), wrap([4, 5, 6]), wrap([7, 8, 9])];
      expect(unwrap(roundRobin(...sources))).toEqual([1, 4, 7, 2, 5, 8, 3, 6, 9]);
    });
  });

  describe('when sources are different sizes', () => {
    it('yields all values collated', () => {
      const sources = [wrap([]), wrap([1, 2, 3]), wrap([4])];
      expect(unwrap(roundRobin(...sources))).toEqual([1, 4, 2, 3]);
    });
  });

  describe('when the interleave generator does not yield all values', () => {
    it('closes the source iterables', () => {
      const incompleteInterleave = interleave(function* () {});

      expect(unwrap(incompleteInterleave(wrap([1, 2, 3]), wrap([4, 5, 6])))).toEqual([]);
    });
  });

  describe('when the interleave is terminated abruptly', () => {
    // eslint-disable-next-line jest/expect-expect
    it('calls return on the strategy', () => {
      const iter = interleave(
        (...args: any) => wrap(anyType(roundRobinStrategy)(...args)),
        {},
        wrap([1, 2, 3]),
        wrap([4, 5, 6]),
      );
      iter.next();
      iter.next();
      iter.return();
    });
  });

  it('can be passed options for the generator', () => {
    const options = {};

    expect.assertions(1);
    unwrap(
      interleave(
        function* (o: {}): Iterable<any> {
          expect(o).toBe(options);
        },
        options,
        null,
      ),
    );
  });

  describe('the summary', () => {
    it('summary.value is a buffer', () => {
      const concat = interleave(function* (
        _: Record<string, any>,
        all: Peekerator<Peekerator<number>>,
      ) {
        let first = true;
        while (!all.done) {
          if (first) {
            // eslint-disable-next-line jest/no-conditional-expect
            expect(() => all.advance()).toThrowErrorMatchingSnapshot();
          }
          expect(all.current.done).toBe(all.done);
          expect(all.current.value).toBe(all.value);
          const { value: buffer, index } = all;
          yield { value: buffer.value, index };
          buffer.advance();
          first = false;
        }
      });

      expect(unwrapDeep(concat(wrap([1]), wrap([2]), wrap([3])))).toEqual([
        {
          index: 0,
          value: 1,
        },
        {
          index: 1,
          value: 2,
        },
        {
          index: 2,
          value: 3,
        },
      ]);
    });
  });
});
