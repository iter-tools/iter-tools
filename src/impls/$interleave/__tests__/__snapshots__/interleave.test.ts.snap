// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interleave the summary summary.value is a buffer 1`] = `"advance() is not supported on an interleave summary"`;
